### PR TITLE
Implement cocharacter and primitive Eulerian polynomials for hyperplane arrangements

### PR DESCRIPTION
**Pull request by tscrim**
### :books: Description

As defined in a recent paper https://arxiv.org/abs/2306.15556.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by tscrim** [comment]
@saliola I implemented this following your paper,  but I don't get the positivity for simplicial arrangements you claim:
```
sage: H = hyperplane_arrangements.braid(4)
sage: H.is_simplicial()
True
sage: H.primitive_eulerian_polynomial()
6*z^3 - 7*z^2 + 2*z
```
Did I make a mistake my implementation?

**Comment by saliola** [comment]
Check the minimal element. We use the intersection of all the hyperplanes as the minimal element, which I think this is opposite to Sage's convention (inclusion vs reverse inclusion).

**Comment by tscrim** [comment]
I had thought of that. Sage is also using the 0 dimensional subspace as the minimal element. Do you have a nonsymmteric poset on hand to check?

**Comment by josebastidasolaya** [comment]
It seems that in the example, the minimal element is the ambient space--expressed as the sum of `p` (the origin) and `W` (the ambient space itself).

**Comment by tscrim** [comment]
Ah, you’re right, I misread it. I will fix it on Monday.  Thank you.

**Comment by tscrim** [comment]
Thank you. I have fixed it and added a lot more examples.

**Comment by tscrim** [comment]
Thank you. @josebastidasolaya I will interpret that as a positive review. Although we now need @saliola to approve the PR request since permissions are now much more restrictive in this new workflow...

**Comment by fchapoton** [comment]
doc does not build

`[sagemath_doc_html-none] OSError: /sage/src/sage/geometry/hyperplane_arrangement/arrangement.py:docstring of sage.geometry.hyperplane_arrangement.arrangement.HyperplaneArrangementElement.primitive_eulerian_polynomial:3: WARNING: citation not found: BHS2023`

**Comment by vbraun** [comment]
pdf docs don't build

**Comment by tscrim** [comment]
@fchapoton Thank you for fixing it. Sorry I couldn't get to it sooner; I was traveling to the US (for FPSAC this week).

**Comment by vbraun** [comment]
```
[sagemath_doc_pdf-none] ! Undefined control sequence.
[sagemath_doc_pdf-none] <argument> ... \in L} |\mu (B,X)| (z - 1)^{\codim 
[sagemath_doc_pdf-none]                                                   X},\end {split}
[sagemath_doc_pdf-none] l.2887 ...\mu(B,X)| (z - 1)^{\codim X},\end{split}
```

**Comment by tscrim** [comment]
Sorry about that. This will fix that issue. I also added one additional parenthetical comment to the documentation specifying that the minimal element is the 0 dimensional subspace (which is opposite of Sage’s convention of being the entire space).

**Comment by tscrim** [comment]
@saliola, @kwankyu, @fchapoton  Could someone approve the PR since it is a positive review?

**Review by kwankyu — APPROVED** [review]
LGTM.

File: src/sage/geometry/hyperplane_arrangement/arrangement.py [hunk 1]
```diff
@@ -975,6 +975,177 @@ def poincare_polynomial(self):
...
+        r"""
+        Return the cocharacteristic polynomial of ``self``.
+
+        The cocharacteristic polynomial of a hyperplane arrangement `A`
+        is defined by
+
+        .. MATH::
+
+            \Psi_A(z) := sum_{X \in L} |\mu(B,X)| z^{\dim X},
+
```
Comment (kwankyu) [inline src/sage/geometry/hyperplane_arrangement/arrangement.py:989]: `\sum`

File: src/sage/geometry/hyperplane_arrangement/arrangement.py [hunk 2]
```diff
@@ -975,6 +975,177 @@ def poincare_polynomial(self):
...
+        r"""
+        Return the primitive Eulerian polynomial of ``self``.
+
+        The primitive Eulerian polynomial of a hyperplane arrangement `A`
+        is defined [BHS2023]_ by
+
+        .. MATH::
+
+            P_A(z) := sum_{X \in L} |\mu(B,X)| (z - 1)^{\mathrm{codim} X},
+
```
Comment (kwankyu) [inline src/sage/geometry/hyperplane_arrangement/arrangement.py:1039]: `\sum`

**Comment by kwankyu** [comment]
> Thank you. @josebastidasolaya I will interpret that as a positive review. 

I don't know what "that" is, which was enough for you to "interpret that" as a positive review. I think you should have requested it explicitly from him. 

As he did not express objection to your interpretation and now I approved this PR, I would not bother about it anymore.

File: src/sage/geometry/hyperplane_arrangement/arrangement.py [hunk 3]
```diff
@@ -975,6 +975,177 @@ def poincare_polynomial(self):
...
+        `B_k` for `k = 2,3,4,5`::
+
+            sage: B = [hyperplane_arrangements.braid(k) for k in range(2,6)]
+            sage: all(H.is_simplicial() for H in B)
+            True
+            sage: all(c > 0 for H in B for c in H.primitive_eulerian_polynomial().coefficients())
+            True
+
+        We verify Example 9.4 in [BHS2023]_ showing a hyperplane arrangemen
+        whose primitive Eulerian polynomial does not have real roots (in
```
Comment (kwankyu) [inline src/sage/geometry/hyperplane_arrangement/arrangement.py:1112]: `arrangement`

**Comment by kwankyu** [comment]
Though this PR has no problem code-wise, I have a little concern about its content:

The PR implements two concepts defined in a paper published in arXiv very recently (June 2023?). First I admit that I know nothing about the research field of hyperplane arrangements. But it seems to me that the two concepts, as defined in a very recent paper, have not yet been known widely in the field. Then it seems too early to get their implementations into sage. I think that what is implemented in sage should be something of general interest in the field, not only by a few readers and authors of a paper kept in arXiv for a few months.

I may be wrong as I don't know the field. Then please correct me.

**Comment by tscrim** [comment]
> Though this PR has no problem code-wise, I have a little concern about its content:
> 
> The PR implements two concepts defined in a paper published in arXiv very recently (June 2023?). First I admit that I know nothing about the research field of hyperplane arrangements. But it seems to me that the two concepts, as defined in a very recent paper, have not yet been known widely in the field. Then it seems too early to get their implementations into sage. I think that what is implemented in sage should be something of general interest in the field, not only by a few readers and authors of a paper kept in arXiv for a few months.

IMO, that is a very arbitrary/ill-defined standard (I would also say it has already been a few months too) and it is contrary to Sage's goal of being a computational tool for experimental mathematics through sharing code. I don't see an argument in your post for why we should not include new and/or not-widely-known (in some definition) algorithms/computations/objects/etc. Yes, it can be difficult to change nomenclature (such as if a conflict arises), but we can still change method names if a need arises later.

**Comment by kwankyu** [comment]
I did not attempt to define a standard of what can or cannot be included into Sage. I doubt who could.

In practice, it is all judged by the authors and reviewers of a PR. As a reviewer, I made my judgement based on the circumstances, which is not of much weight of course as I don't know the field. If another reviewer who knows the field think different and accepts the PR, I would not object.

For my judgement, I would not accept implementations of concepts not of general interest in the field.

**Review by fchapoton — APPROVED** [review]
ok

**Pull request merged**